File: tests/test_builder_casts.cpp
/**
 * Test: test_builder_casts
 * Tests LLVM Builder cast operations
 *
 * LLVM-C APIs covered:
 * - LLVMBuildTrunc(), LLVMBuildZExt(), LLVMBuildSExt()
 * - LLVMBuildFPToUI(), LLVMBuildFPToSI()
 * - LLVMBuildUIToFP(), LLVMBuildSIToFP()
 * - LLVMBuildFPTrunc(), LLVMBuildFPExt()
 * - LLVMBuildPtrToInt(), LLVMBuildIntToPtr()
 * - LLVMBuildBitCast()
 * - LLVMBuildIntCast2()
 */

#include <cstdio>
#include <llvm-c/Analysis.h>
#include <llvm-c/Core.h>

int main() {
  LLVMContextRef ctx = LLVMContextCreate();
  LLVMModuleRef mod =
      LLVMModuleCreateWithNameInContext("test_builder_casts", ctx);

  LLVMTypeRef i8 = LLVMInt8TypeInContext(ctx);
  LLVMTypeRef i16 = LLVMInt16TypeInContext(ctx);
  LLVMTypeRef i32 = LLVMInt32TypeInContext(ctx);
  LLVMTypeRef i64 = LLVMInt64TypeInContext(ctx);
  LLVMTypeRef f32 = LLVMFloatTypeInContext(ctx);
  LLVMTypeRef f64 = LLVMDoubleTypeInContext(ctx);
  LLVMTypeRef ptr = LLVMPointerTypeInContext(ctx, 0);
  LLVMTypeRef void_ty = LLVMVoidTypeInContext(ctx);

  LLVMBuilderRef builder = LLVMCreateBuilderInContext(ctx);

  // ==========================================
  // Function 1: Integer casts
  // ==========================================
  LLVMTypeRef int_cast_params[] = {i64};
  LLVMTypeRef int_cast_ty = LLVMFunctionType(i8, int_cast_params, 1, 0);
  LLVMValueRef int_cast_func =
      LLVMAddFunction(mod, "integer_casts", int_cast_ty);
  LLVMValueRef i64_val = LLVMGetParam(int_cast_func, 0);
  LLVMSetValueName2(i64_val, "val", 3);

  LLVMBasicBlockRef int_entry =
      LLVMAppendBasicBlockInContext(ctx, int_cast_func, "entry");
  LLVMPositionBuilderAtEnd(builder, int_entry);

  // Truncate i64 -> i32 -> i16 -> i8
  LLVMValueRef trunc_32 = LLVMBuildTrunc(builder, i64_val, i32, "trunc_32");
  LLVMValueRef trunc_16 = LLVMBuildTrunc(builder, trunc_32, i16, "trunc_16");
  LLVMValueRef trunc_8 = LLVMBuildTrunc(builder, trunc_16, i8, "trunc_8");

  // Zero extend i8 -> i16 -> i32 -> i64
  LLVMValueRef zext_16 = LLVMBuildZExt(builder, trunc_8, i16, "zext_16");
  LLVMValueRef zext_32 = LLVMBuildZExt(builder, zext_16, i32, "zext_32");
  LLVMValueRef zext_64 = LLVMBuildZExt(builder, zext_32, i64, "zext_64");

  // Sign extend i8 -> i16 -> i32 -> i64
  LLVMValueRef sext_16 = LLVMBuildSExt(builder, trunc_8, i16, "sext_16");
  LLVMValueRef sext_32 = LLVMBuildSExt(builder, sext_16, i32, "sext_32");
  LLVMValueRef sext_64 = LLVMBuildSExt(builder, sext_32, i64, "sext_64");

  // IntCast2 (auto-selects trunc/zext/sext)
  LLVMValueRef intcast_unsigned =
      LLVMBuildIntCast2(builder, i64_val, i32, 0, "intcast_unsigned");
  LLVMValueRef intcast_signed =
      LLVMBuildIntCast2(builder, trunc_8, i32, 1, "intcast_signed");

  LLVMBuildRet(builder, trunc_8);

  // ==========================================
  // Function 2: Float casts
  // ==========================================
  LLVMTypeRef fp_cast_params[] = {f64};
  LLVMTypeRef fp_cast_ty = LLVMFunctionType(f32, fp_cast_params, 1, 0);
  LLVMValueRef fp_cast_func = LLVMAddFunction(mod, "float_casts", fp_cast_ty);
  LLVMValueRef f64_val = LLVMGetParam(fp_cast_func, 0);
  LLVMSetValueName2(f64_val, "val", 3);

  LLVMBasicBlockRef fp_entry =
      LLVMAppendBasicBlockInContext(ctx, fp_cast_func, "entry");
  LLVMPositionBuilderAtEnd(builder, fp_entry);

  // FP truncate f64 -> f32
  LLVMValueRef fptrunc = LLVMBuildFPTrunc(builder, f64_val, f32, "fptrunc");

  // FP extend f32 -> f64
  LLVMValueRef fpext = LLVMBuildFPExt(builder, fptrunc, f64, "fpext");

  LLVMBuildRet(builder, fptrunc);

  // ==========================================
  // Function 3: Int <-> Float casts
  // ==========================================
  LLVMTypeRef mixed_params[] = {i32, f64};
  LLVMTypeRef mixed_ty = LLVMFunctionType(void_ty, mixed_params, 2, 0);
  LLVMValueRef mixed_func = LLVMAddFunction(mod, "int_float_casts", mixed_ty);
  LLVMValueRef int_param = LLVMGetParam(mixed_func, 0);
  LLVMValueRef fp_param = LLVMGetParam(mixed_func, 1);
  LLVMSetValueName2(int_param, "i", 1);
  LLVMSetValueName2(fp_param, "f", 1);

  LLVMBasicBlockRef mixed_entry =
      LLVMAppendBasicBlockInContext(ctx, mixed_func, "entry");
  LLVMPositionBuilderAtEnd(builder, mixed_entry);

  // Int -> Float
  LLVMValueRef uitofp = LLVMBuildUIToFP(builder, int_param, f64, "uitofp");
  LLVMValueRef sitofp = LLVMBuildSIToFP(builder, int_param, f64, "sitofp");

  // Float -> Int
  LLVMValueRef fptoui = LLVMBuildFPToUI(builder, fp_param, i32, "fptoui");
  LLVMValueRef fptosi = LLVMBuildFPToSI(builder, fp_param, i32, "fptosi");

  LLVMBuildRetVoid(builder);

  // ==========================================
  // Function 4: Pointer casts
  // ==========================================
  LLVMTypeRef ptr_params[] = {ptr, i64};
  LLVMTypeRef ptr_ty = LLVMFunctionType(void_ty, ptr_params, 2, 0);
  LLVMValueRef ptr_func = LLVMAddFunction(mod, "pointer_casts", ptr_ty);
  LLVMValueRef ptr_param = LLVMGetParam(ptr_func, 0);
  LLVMValueRef int_for_ptr = LLVMGetParam(ptr_func, 1);
  LLVMSetValueName2(ptr_param, "p", 1);
  LLVMSetValueName2(int_for_ptr, "addr", 4);

  LLVMBasicBlockRef ptr_entry =
      LLVMAppendBasicBlockInContext(ctx, ptr_func, "entry");
  LLVMPositionBuilderAtEnd(builder, ptr_entry);

  // Pointer -> Int
  LLVMValueRef ptrtoint =
      LLVMBuildPtrToInt(builder, ptr_param, i64, "ptrtoint");

  // Int -> Pointer
  LLVMValueRef inttoptr =
      LLVMBuildIntToPtr(builder, int_for_ptr, ptr, "inttoptr");

  LLVMBuildRetVoid(builder);

  // ==========================================
  // Function 5: Bitcast
  // ==========================================
  LLVMTypeRef vec_i32 = LLVMVectorType(i32, 4);
  LLVMTypeRef vec_f32 = LLVMVectorType(f32, 4);

  LLVMTypeRef bitcast_params[] = {vec_i32};
  LLVMTypeRef bitcast_ty = LLVMFunctionType(vec_f32, bitcast_params, 1, 0);
  LLVMValueRef bitcast_func =
      LLVMAddFunction(mod, "bitcast_example", bitcast_ty);
  LLVMValueRef vec_param = LLVMGetParam(bitcast_func, 0);
  LLVMSetValueName2(vec_param, "v", 1);

  LLVMBasicBlockRef bitcast_entry =
      LLVMAppendBasicBlockInContext(ctx, bitcast_func, "entry");
  LLVMPositionBuilderAtEnd(builder, bitcast_entry);

  LLVMValueRef bitcast =
      LLVMBuildBitCast(builder, vec_param, vec_f32, "bitcast");
  LLVMBuildRet(builder, bitcast);

  LLVMDisposeBuilder(builder);

  // Verify module
  char *error = nullptr;
  if (LLVMVerifyModule(mod, LLVMReturnStatusAction, &error)) {
    fprintf(stderr, "; Verification failed: %s\n", error);
    LLVMDisposeMessage(error);
    LLVMDisposeModule(mod);
    LLVMContextDispose(ctx);
    return 1;
  }
  LLVMDisposeMessage(error);

  // Print diagnostic comments
  printf("; Test: test_builder_casts\n");
  printf(";\n");
  printf("; Cast operations demonstrated:\n");
  printf(";   Integer: trunc, zext, sext, intcast2\n");
  printf(";   Float: fptrunc, fpext\n");
  printf(";   Int<->Float: uitofp, sitofp, fptoui, fptosi\n");
  printf(";   Pointer: ptrtoint, inttoptr\n");
  printf(";   Reinterpret: bitcast\n");
  printf("\n");

  // Print module IR
  char *ir = LLVMPrintModuleToString(mod);
  printf("%s", ir);
  LLVMDisposeMessage(ir);

  // Cleanup
  LLVMDisposeModule(mod);
  LLVMContextDispose(ctx);

  return 0;
}

File: tests/test_builder_casts.py
#!/usr/bin/env -S uv run
"""
Test: test_builder_casts
Tests LLVM Builder cast operations

Python equivalent of tests/test_builder_casts.cpp
Must produce identical output to the C++ version.
"""

import llvm


def main():
    with llvm.create_context() as ctx:
        with ctx.create_module("test_builder_casts") as mod:
            i8 = ctx.types.i8
            i16 = ctx.types.i16
            i32 = ctx.types.i32
            i64 = ctx.types.i64
            f32 = ctx.types.f32
            f64 = ctx.types.f64
            ptr = ctx.types.ptr
            void_ty = ctx.types.void

            assert i8 == ctx.types.int_n(8), "oh nein"

            # ==========================================
            # Function 1: Integer casts
            # ==========================================
            int_cast_ty = ctx.types.function(i8, [i64])
            int_cast_func = mod.add_function("integer_casts", int_cast_ty)
            i64_val = int_cast_func.get_param(0)
            i64_val.name = "val"

            int_entry = int_cast_func.append_basic_block("entry")

            with int_entry.create_builder() as builder:
                # Truncate i64 -> i32 -> i16 -> i8
                trunc_32 = builder.trunc(i64_val, i32, "trunc_32")
                trunc_16 = builder.trunc(trunc_32, i16, "trunc_16")
                trunc_8 = builder.trunc(trunc_16, i8, "trunc_8")

                # Zero extend i8 -> i16 -> i32 -> i64
                zext_16 = builder.zext(trunc_8, i16, "zext_16")
                zext_32 = builder.zext(zext_16, i32, "zext_32")
                zext_64 = builder.zext(zext_32, i64, "zext_64")

                # Sign extend i8 -> i16 -> i32 -> i64
                sext_16 = builder.sext(trunc_8, i16, "sext_16")
                sext_32 = builder.sext(sext_16, i32, "sext_32")
                sext_64 = builder.sext(sext_32, i64, "sext_64")

                # IntCast2 (auto-selects trunc/zext/sext)
                intcast_unsigned = builder.int_cast2(
                    i64_val, i32, False, "intcast_unsigned"
                )
                intcast_signed = builder.int_cast2(trunc_8, i32, True, "intcast_signed")

                builder.ret(trunc_8)

                # ==========================================
                # Function 2: Float casts
                # ==========================================
                fp_cast_ty = ctx.types.function(f32, [f64])
                fp_cast_func = mod.add_function("float_casts", fp_cast_ty)
                f64_val = fp_cast_func.get_param(0)
                f64_val.name = "val"

                fp_entry = fp_cast_func.append_basic_block("entry")
                builder.position_at_end(fp_entry)

                # FP truncate f64 -> f32
                fptrunc = builder.fptrunc(f64_val, f32, "fptrunc")

                # FP extend f32 -> f64
                fpext = builder.fpext(fptrunc, f64, "fpext")

                builder.ret(fptrunc)

                # ==========================================
                # Function 3: Int <-> Float casts
                # ==========================================
                mixed_ty = ctx.types.function(void_ty, [i32, f64])
                mixed_func = mod.add_function("int_float_casts", mixed_ty)
                int_param = mixed_func.get_param(0)
                fp_param = mixed_func.get_param(1)
                int_param.name = "i"
                fp_param.name = "f"

                mixed_entry = mixed_func.append_basic_block("entry")
                builder.position_at_end(mixed_entry)

                # Int -> Float
                uitofp = builder.uitofp(int_param, f64, "uitofp")
                sitofp = builder.sitofp(int_param, f64, "sitofp")

                # Float -> Int
                fptoui = builder.fptoui(fp_param, i32, "fptoui")
                fptosi = builder.fptosi(fp_param, i32, "fptosi")

                builder.ret_void()

                # ==========================================
                # Function 4: Pointer casts
                # ==========================================
                ptr_ty = ctx.types.function(void_ty, [ptr, i64])
                ptr_func = mod.add_function("pointer_casts", ptr_ty)
                ptr_param = ptr_func.get_param(0)
                int_for_ptr = ptr_func.get_param(1)
                ptr_param.name = "p"
                int_for_ptr.name = "addr"

                ptr_entry = ptr_func.append_basic_block("entry")
                builder.position_at_end(ptr_entry)

                # Pointer -> Int
                ptrtoint = builder.ptrtoint(ptr_param, i64, "ptrtoint")

                # Int -> Pointer
                inttoptr = builder.inttoptr(int_for_ptr, ptr, "inttoptr")

                builder.ret_void()

                # ==========================================
                # Function 5: Bitcast
                # ==========================================
                vec_i32 = i32.vector(4)
                vec_f32 = f32.vector(4)

                bitcast_ty = ctx.types.function(vec_f32, [vec_i32])
                bitcast_func = mod.add_function("bitcast_example", bitcast_ty)
                vec_param = bitcast_func.get_param(0)
                vec_param.name = "v"

                bitcast_entry = bitcast_func.append_basic_block("entry")
                builder.position_at_end(bitcast_entry)

                bitcast = builder.bitcast(vec_param, vec_f32, "bitcast")
                builder.ret(bitcast)

            # Verify module
            if not mod.verify():
                print(f"; Verification failed: {mod.get_verification_error()}")
                return 1

            # Print diagnostic comments
            print("; Test: test_builder_casts")
            print(";")
            print("; Cast operations demonstrated:")
            print(";   Integer: trunc, zext, sext, intcast2")
            print(";   Float: fptrunc, fpext")
            print(";   Int<->Float: uitofp, sitofp, fptoui, fptosi")
            print(";   Pointer: ptrtoint, inttoptr")
            print(";   Reinterpret: bitcast")
            print()

            # Print module IR
            print(mod.to_string(), end="")

    return 0


if __name__ == "__main__":
    exit(main())
